PowerShell Environment Variables Module - closes #155 (#176)
bob-crutchley/notes: topics/powershell/modules/environment-variables/README.md

		- [System Level Example](#system-level-example)
- [Unassign Environment Variables](#unassign-environment-variables)
	- [Unassign Process Level Environment Variable](#unassign-process-level-environment-variable)
	- [Unassign User Level Environment Variable](#unassign-user-level-environment-variable)
	- [Unassign System Level Environment Variable](#unassign-system-level-environment-variable)
- [Tasks](#tasks)
	- [Process Level Assignment](#process-level-assignment-1)
	- [User & System Level Assignment](#user--system-level-assignment-1)
		- [User Level](#user-level)
		- [System Level](#system-level)
	- [Viewing System & User Level Environment Variables on Windows](#viewing-system--user-level-environment-variables-on-windows)
	- [Unassign the Environment Variables](#unassign-the-environment-variables)

<!--TOC_END-->
## Overview
PowerShell has the ability to set and access system environment variables.
Environment variables are just like normal variables, except they are configured outside of applications.
This gives us the ability to program applications more generically as opposed to hard-coding the values that we want into the program itself.
## Access Types

> Environment variables differ from regular variables, because they are not subject to scope like regular variables are.

Reviewer: ```suggestion
Environment variables differ from regular variables, because they are not subjected to scope like regular variables are.
```

> There are however different types which you can set an environment variable which affects access to it.

Reviewer: ```suggestion
There are however, different access types which you can set on an environment variable which affects access to it.
```

|Access Type|Description|
|-----------|-----------|
|Process|The current PowerShell session and any applications started within it will be able to access the variable. Environment variables at this level can be thought of as temporary, when the process stops or the system restarts, the variable will no longer exist.|
|User|Any applications being run by the same system user will be able to access the variable. The variable is persisted and will still be accessible after a reboot.|
|System|Access is allowed across the entire system, the variable will remain after a restart. Only administrators can modify these variables.|

### Which Access Type and When?

> This is can get fairly subjective but just from consiering the scopes for the variables it isn't too hard to figure out which one you need, especially when you consider; does it need to be accessible at this scope?

Reviewer: ```suggestion
Deciding on which access type to use can get fairly subjective, but just from considering the scopes for the variables it isn't too hard to figure it out. Try answering the following question in order to decide what scope to use: does it need to be accessible at this scope?
```

#### Example Scenario
Lets consider the following scenario as an example:
- Deploying an application server that needs a database connection passed in as an environment variable

> - Application server executable is run by the `app` user, 

Reviewer: ```suggestion
- Application server executable is run by the `app` user
```

> - The application is executed as a service (background process not started by a PowerShell session)

Reviewer: ```suggestion
- The application is executed as a service (a background process that is not started by a PowerShell session)
```

From this information we can see that a System environment variable is not a good option here, because only the `app` user needs access to the value stored in that environment variable.
If every user on the system needed to be able to access the same variable then that would be a better scenario to have the envinronment variable as a System access type.

We can't really use a Process level environment variable because the application is executed as a service, so not executed by a PowerShell session.

This leaves us with the User level which will mean that any process running as the `app` user will have access to this variable.
This will work with the application being executed as a service.

## Accessing an Environment Variable
Environment variables can be accessed in PowerShell by use `$env:` and the name of the environment variable:
```powershell

> echo $env:HOMEPATH

Reviewer: is there a difference how you access the different scope variables?

Author: No, they are all accessed in the same way.
The scope will only change whether they are accessible or not.
I've added a sentence to explain this now.

```
## Assigning Environment Variables
There are a couple of ways to assign an environment variable depend on what access type is going to be needed.
### Process Level Assignment
You can use `$env` to assign a process level environment variable:
```powershell
$env:TEST_VARIABLE = “test variable value”
```
### User & System Level Assignment
User and System level environment variable assigment will need to be configured by leveraging the .NET framework that PowerShell has access to.
This works by accessing the `SetEnvironmentVariable` method from the `System.Environment` class.
We can pass this method the variable `name`, `value` and the `type`.
#### User Level Example
```powershell
[System.Environment]::SetEnvironmentVariable("TEST_VARIABLE", "test variable", "User")
```
#### System Level Example
```powershell
System-wide environment variables can be provided by using the `Machine` type, which seems a bit ambiguous considering its for a System environment variable.

> Remember, when you are setting a System environment variable you will need have a PowerShell session opened as an administrator on Windows.

Reviewer: Could add a section where you show error where you tried to set it not as an administrator, and tell to run it as admin

Author: Good idea, I've added a suggestion for this just before the example.

```
### Unassign User Level Environment Variable
```powershell
[System.Environment]::SetEnvironmentVariable("TEST_VARIABLE", $null, "User")
```
### Unassign System Level Environment Variable
```powershell
[System.Environment]::SetEnvironmentVariable("TEST_VARIABLE", $null, "Machine")
```

## Tasks
Although PowerShell can be used on other systems like Linux, these exercises are best suited on a Windows machine so that you can see the the different access types working correctly.
You will find on Linux, only process environment variables seem to work correctly.

### Process Level Assignment
Lets set a process level environment variable called `TEST_PROCESS_VARIABLE` with a value of `test process variable`:
```powershell
$env:TEST_PROCESS_VARIABLE = “test process variable”
```
We can now check that the variable has been assigned correctly by accessing it:
```powershell
echo $env:TEST_PROCESS_VARIABLE
```
### User & System Level Assignment
Now we can assign simliar variables but on a `User` and `System` level by using PowerShell's .NET capabilities.
#### User Level
```powershell
[System.Environment]::SetEnvironmentVariable("TEST_USER_VARIABLE", "test user variable", "User")

> echo $env:TEST_USER_VARIABLE

Reviewer: doesn't print anything

Author: So it looks like the issue here is that the PowerShell session needs to be restarted before the variables will be accessible.
I've now made some mentions of this throughout the handout and added reminders on the task itself to restart the PowerShell session after running the command.

```
#### System Level
```powershell
[System.Environment]::SetEnvironmentVariable("TEST_SYSTEM_VARIABLE", "test system variable", "Machine")

> echo $env:TEST_SYSTEM_VARIABLE

Reviewer: doesn't print anything

Author: Same issue as this: https://github.com/bob-crutchley/notes/pull/176/files#r331992559

```powershell
echo $env:HOMEPATH
```
All the types of environment variables we have discussed are all accessed in the same way.
The access level will only change whether they are accessible or not.

When setting an environment variable that is either for `User` or `System` level, you will need to restart the PowerShell session before you will be able to access it.
## Assigning Environment Variables
There are a couple of ways to assign an environment variable depend on what access type is going to be needed.
### Process Level Assignment
You can use `$env` to assign a process level environment variable:
```powershell
$env:TEST_VARIABLE = “test variable value”
```
### User & System Level Assignment
User and System level environment variable assigment will need to be configured by leveraging the .NET framework that PowerShell has access to.
This works by accessing the `SetEnvironmentVariable` method from the `System.Environment` class.
We can pass this method the variable `name`, `value` and the `type`.
#### User Level Example
```powershell
[System.Environment]::SetEnvironmentVariable("TEST_VARIABLE", "test variable", "User")
```
#### System Level Example

> System-wide environment variables can be provided by using the `Machine` type, which seems a bit ambiguous considering its for a System environment variable.

Reviewer 2: ```suggestion
System-wide environment variables can be provided by using the `Machine` type, which seems a bit ambiguous considering it's for a System environment variable.
```

> Remember, when you are setting a System environment variable you will need have a PowerShell session opened as an administrator on Windows.

Reviewer 2: ```suggestion
Remember, when you are setting a System environment variable, you will need have a PowerShell session opened as an administrator on Windows.
```

You could try running the example shown here in a non-administritive PowerShell session first to see the error, then try running it in an administrivtive PowerShell window:
```powershell
[System.Environment]::SetEnvironmentVariable("TEST_VARIABLE", "test variable", "Machine")
```
## Unassign Environment Variables

> Variables can be effectively unassigned just by setting them to be `$null`.

Reviewer 2: ```suggestion
Variables can be unassigned by setting them to be `$null`.
```

> So to unassign a variable, you just assign `$null` to it.

Reviewer 2: ```suggestion
So, to unassign a variable, you just assign `$null` to it.
```

Some examples have been included in this section for extra clarification.
### Unassign Process Level Environment Variable
```powershell
$env:TEST_VARIABLE = $null
```
### Unassign User Level Environment Variable
```powershell
[System.Environment]::SetEnvironmentVariable("TEST_VARIABLE", $null, "User")
```
### Unassign System Level Environment Variable
```powershell
[System.Environment]::SetEnvironmentVariable("TEST_VARIABLE", $null, "Machine")
```

## Tasks

> Although PowerShell can be used on other systems like Linux, these exercises are best suited on a Windows machine so that you can see the the different access types working correctly.

Reviewer 2: ```suggestion
Although PowerShell can be used on other systems like Linux, these exercises are best suited on a Windows machine. This is so that you can see the the different access types working correctly.
```

> You will find on Linux, only process environment variables seem to work correctly.

Reviewer 2: ```suggestion
You will find that only process environment variables seem to work correctly on Linux.
```

### Process Level Assignment

> Lets set a process level environment variable called `TEST_PROCESS_VARIABLE` with a value of `test process variable`:

Reviewer 2: ```suggestion
Let's set a process level environment variable called `TEST_PROCESS_VARIABLE`, with a value of `test process variable`:
```

```powershell
$env:TEST_PROCESS_VARIABLE = “test process variable”
```
We can now check that the variable has been assigned correctly by accessing it:
```powershell
echo $env:TEST_PROCESS_VARIABLE
```
### User & System Level Assignment

> Now we can assign simliar variables but on a `User` and `System` level by using PowerShell's .NET capabilities.

Reviewer 2: ```suggestion
Now we can assign simliar variables, but on a `User` and `System` level, by using PowerShell's .NET capabilities.
```

#### User Level
```powershell
[System.Environment]::SetEnvironmentVariable("TEST_USER_VARIABLE", "test user variable", "User")
# restart your powershell session
echo $env:TEST_USER_VARIABLE
```
#### System Level
```powershell
[System.Environment]::SetEnvironmentVariable("TEST_SYSTEM_VARIABLE", "test system variable", "Machine")
# restart your powershell session
echo $env:TEST_SYSTEM_VARIABLE
```
### Viewing System & User Level Environment Variables on Windows
Specifically for Windows, we are able to view the System and User environment variables that have been assigned.

A common way to open the Environment Variables window is by navigating through these steps:
- Right click on the Start Menu 
- Select Control Panel
- Select System
- Select Advanced System Settings
- Select Environment Variables

> The fastest way we can get to the Environment Variables open for the purpose this exercise is by using the Window Run application to open it:

Reviewer 2: ```suggestion
The fastest way we can get to the Environment Variables open, for the purpose of this exercise, is by using the Window Run application to open it:
```
